File: PDFaCodigoJavaCSharpRustPython/tarea_pa_Avanzada_RELOG/PythonCode/Stopwatch.py
import tkinter as tk
import time

class Stopwatch(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("Stopwatch Application")
        
        # Al cerrar la ventana
        self.protocol("WM_DELETE_WINDOW", self.exitForm)

        # declare class level variables
        self.startTime = 0
        self.stopTime = 0
        self.elapsedTime = 0.0

        # declare controls used
        self.startButton = tk.Button(self, text="Start Timing", command=self.startButtonActionPerformed)
        self.stopButton = tk.Button(self, text="Stop Timing", command=self.stopButtonActionPerformed)
        self.exitButton = tk.Button(self, text="Exit", command=self.exitButtonActionPerformed)
        
        self.startLabel = tk.Label(self, text="Start Time")
        self.stopLabel = tk.Label(self, text="Stop Time")
        self.elapsedLabel = tk.Label(self, text="Elapsed Time (sec)")
        
        self.startTextField = tk.Entry(self, width=15)
        self.stopTextField = tk.Entry(self, width=15)
        self.elapsedTextField = tk.Entry(self, width=15)

        # add controls (Equivalente a GridBagConstraints)
        self.startButton.grid(row=0, column=0, padx=5, pady=5)
        self.stopButton.grid(row=1, column=0, padx=5, pady=5)
        self.exitButton.grid(row=2, column=0, padx=5, pady=5)

        self.startLabel.grid(row=0, column=1, padx=5, pady=5)
        self.stopLabel.grid(row=1, column=1, padx=5, pady=5)
        self.elapsedLabel.grid(row=2, column=1, padx=5, pady=5)

        self.startTextField.grid(row=0, column=2, padx=5, pady=5)
        self.stopTextField.grid(row=1, column=2, padx=5, pady=5)
        self.elapsedTextField.grid(row=2, column=2, padx=5, pady=5)

    def startButtonActionPerformed(self):
        # click of start timing button
        self.startTime = int(time.time() * 1000)
        
        self.startTextField.delete(0, tk.END)
        self.startTextField.insert(0, str(self.startTime))
        
        self.stopTextField.delete(0, tk.END)
        self.elapsedTextField.delete(0, tk.END)

    def stopButtonActionPerformed(self):
        # click of stop timing button
        self.stopTime = int(time.time() * 1000)
        
        self.stopTextField.delete(0, tk.END)
        self.stopTextField.insert(0, str(self.stopTime))
        
        self.elapsedTime = (self.stopTime - self.startTime) / 1000.0
        
        self.elapsedTextField.delete(0, tk.END)
        self.elapsedTextField.insert(0, str(self.elapsedTime))

    def exitButtonActionPerformed(self):
        self.destroy()

    def exitForm(self):
        self.destroy()

if __name__ == "__main__":
    app = Stopwatch()
    app.mainloop()

File: PDFaCodigoJavaCSharpRustPython/tarea_pa_Avanzada_RELOG/CSharpcode/Stopwatch.cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace CSharpcode {
    public class Stopwatch : Window {
        // declare controls used
        Button startButton = new Button();
        Button stopButton = new Button();
        Button exitButton = new Button();
        Label startLabel = new Label();
        Label stopLabel = new Label();
        Label elapsedLabel = new Label();
        TextBox startTextField = new TextBox();
        TextBox stopTextField = new TextBox();
        TextBox elapsedTextField = new TextBox();

        // declare class level variables
        long startTime;
        long stopTime;
        double elapsedTime;

        public Stopwatch() {
            Title = "Stopwatch Application";
            Width = 400;
            Height = 200;
            Closed += exitForm; // addWindowListener equivalente

            // Configurar el GridLayout
            Grid grid = new Grid();
            grid.Margin = new Thickness(10);
            for (int i = 0; i < 3; i++) {
                grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
                grid.RowDefinitions.Add(new RowDefinition { Height = new GridLength(1, GridUnitType.Star) });
            }

            // Configurar y añadir controles
            startButton.Content = "Start Timing";
            startButton.Click += startButtonActionPerformed;
            Grid.SetColumn(startButton, 0); Grid.SetRow(startButton, 0);
            grid.Children.Add(startButton);

            stopButton.Content = "Stop Timing";
            stopButton.Click += stopButtonActionPerformed;
            Grid.SetColumn(stopButton, 0); Grid.SetRow(stopButton, 1);
            grid.Children.Add(stopButton);

            exitButton.Content = "Exit";
            exitButton.Click += exitButtonActionPerformed;
            Grid.SetColumn(exitButton, 0); Grid.SetRow(exitButton, 2);
            grid.Children.Add(exitButton);

            startLabel.Content = "Start Time";
            Grid.SetColumn(startLabel, 1); Grid.SetRow(startLabel, 0);
            grid.Children.Add(startLabel);

            stopLabel.Content = "Stop Time";
            Grid.SetColumn(stopLabel, 1); Grid.SetRow(stopLabel, 1);
            grid.Children.Add(stopLabel);

            elapsedLabel.Content = "Elapsed Time (sec)";
            Grid.SetColumn(elapsedLabel, 1); Grid.SetRow(elapsedLabel, 2);
            grid.Children.Add(elapsedLabel);

            startTextField.Text = "";
            Grid.SetColumn(startTextField, 2); Grid.SetRow(startTextField, 0);
            grid.Children.Add(startTextField);

            stopTextField.Text = "";
            Grid.SetColumn(stopTextField, 2); Grid.SetRow(stopTextField, 1);
            grid.Children.Add(stopTextField);

            elapsedTextField.Text = "";
            Grid.SetColumn(elapsedTextField, 2); Grid.SetRow(elapsedTextField, 2);
            grid.Children.Add(elapsedTextField);

            Content = grid;
        }

        private void startButtonActionPerformed(object? sender, RoutedEventArgs e) {
            startTime = DateTimeOffset.Now.ToUnixTimeMilliseconds();
            startTextField.Text = startTime.ToString();
            stopTextField.Text = "";
            elapsedTextField.Text = "";
        }

        private void stopButtonActionPerformed(object? sender, RoutedEventArgs e) {
            stopTime = DateTimeOffset.Now.ToUnixTimeMilliseconds();
            stopTextField.Text = stopTime.ToString();
            elapsedTime = (stopTime - startTime) / 1000.0;
            elapsedTextField.Text = elapsedTime.ToString();
        }

        private void exitButtonActionPerformed(object? sender, RoutedEventArgs e) {
            System.Windows.Application.Current.Shutdown();
        }

private void exitForm(object? sender, EventArgs e) {
        System.Windows.Application.Current.Shutdown();
    }

    // Pega el método Main AQUÍ, dentro de la clase Stopwatch
    [STAThread]
    public static void Main(string[] args) {
        Application app = new Application();
        Stopwatch window = new Stopwatch();
        app.Run(window);
    }
} // <- Llave que cierra la clase Stopwatch
} // <- Llave que cierra el namespace
